chalk-rust/src/lower/mod.rs

```diff
@@ -521,6 +521,16 @@ impl LowerTraitRef for TraitRef {
                   k.binders.len() + 1, parameters.len())
         }
 
+        for (binder, param) in k.binders.binders.iter().zip(parameters.iter().skip(1)) {
+            match (binder, param) {
+                (&ir::ParameterKind::Ty(_), &ir::ParameterKind::Lifetime(_)) =>
+                    bail!("type parameter to trait has wrong kind: expected type, found lifetime"),
+                (&ir::ParameterKind::Lifetime(_), &ir::ParameterKind::Ty(_)) =>
+                    bail!("type parameter to trait has wrong kind: expected lifetime, found type"),
+                _ => {}
```

Comment:
Can you go ahead and make these matches exhaustive? I would probably write it like this:

```
match (binder, param) {
    (&ir::ParameterKind::Ty(_), &ir::ParameterKind::Ty(_)) => { }
    (&ir::ParameterKind::Lifetime(_), &ir::ParameterKind::Lifetime(_)) => { }
    (&ir::ParameterKind::Ty(_), _) |
    (&ir::ParameterKind::Lifetime(_), _) => {
        bail!("incorrect parameter kind: expected {}, found {}", kind(binder), kind(param))
    }
}
```

where `kind` is some helper function. Note that this library is not (yet, anyway) really intended for *end-users per se*, so error messages don't have to be super-duper. Anyway, exhaustive matches really help when refactoring (i.e, adding a new sort of `ParameterKind`), so I try to avoid `_` (in chalk in particular, though also other places).